src/codegen/datatype/datatype_registry.cc

```diff
@@ -0,0 +1,123 @@
+      // TODO(gus) I cast this down to an int so that it can be automatically
+      // converted to TVMRetValue. This is dumb and should be fixed somehow.
+      *ret = (int)DatatypeRegistry::Global()->GetStorageSize(args[0].operator int());
+    });
+
+void DatatypeRegistry::RegisterDatatype(const std::string& type_name,
+                                        uint8_t type_code,
+                                        size_t storage_size) {
+  code_to_name[type_code] = type_name;
+  name_to_code[type_name] = type_code;
+  code_to_storage_size[type_code] = storage_size;
+}
+
+uint8_t DatatypeRegistry::GetTypeCode(const std::string& type_name) {
+  return name_to_code[type_name];
+}
+
+std::string DatatypeRegistry::GetTypeName(uint8_t type_code) {
+  return code_to_name[type_code];
+}
+
+size_t DatatypeRegistry::GetStorageSize(uint8_t type_code) {
+  return code_to_storage_size[type_code];
+}
+
+const runtime::PackedFunc* GetCastLowerFunc(const std::string& target,
+                                            uint8_t type_code,
+                                            uint8_t src_type_code) {
+  std::ostringstream ss;
+  ss << "tvm.datatypes.lower.";
+  ss << target << ".";
+  ss << "Cast"
+     << ".";
+
+  if (DatatypeRegistry::Global()->DatatypeRegistered(type_code)) {
+    ss << DatatypeRegistry::Global()->GetTypeName(type_code);
+  } else {
+    ss << runtime::TypeCode2Str(type_code);
+  }
+
+  ss << ".";
+
+  if (DatatypeRegistry::Global()->DatatypeRegistered(src_type_code)) {
+    ss << DatatypeRegistry::Global()->GetTypeName(src_type_code);
+  } else {
+    ss << runtime::TypeCode2Str(src_type_code);
+  }
+
+  return runtime::Registry::Get(ss.str());
+}
+
+TVM_REGISTER_GLOBAL("_register_Cast")
+    .set_body([](TVMArgs args, TVMRetValue *rv) {
+      const std::string target = args[0];
+      const std::string type = args[1];
+      const std::string src_type = args[2];
+      const std::string extern_func_name = args[3];
+
+      auto lower_cast_name =
+          "tvm.datatypes.lower." + target + ".Cast." + type + "." + src_type;
+      runtime::Registry::Register(lower_cast_name)
+          .set_body([extern_func_name](TVMArgs args, TVMRetValue *rv) {
+            Expr e = args[0];
+            const ir::Cast *cast = e.as<ir::Cast>();
+            internal_assert(cast);
+            // Custom datatypes should get cast to their underlying storage
+            // type.
+            // TODO(gus) I'm not using the width registered originally; I'm
+            // using the bits() attached to the type (where does this come
+            // from?)
+            auto return_type = DatatypeRegistry::Global()->DatatypeRegistered(
+                                   cast->type.code())
+                                   ? UInt(cast->type.bits())
+                                   : cast->type;
+            *rv = ir::Call::make(return_type, extern_func_name, {cast->value},
+                                 ir::Call::Extern);
```

Comment:
Like the comment below this, here is where we define the signature for external cast calls; could we instead take a pointer to an array? Is doing one element at a time inefficient?

Reply:
Another way to phrase this is: what's the best signature to use to make this as flexible and useful as possible? Can we support multiple possible signatures?

Reply:
tq: this is fine for now, and is actually probably the most efficient (if calls can be inlined/if operations can be done in registers instead of going to memory).